fuse_core/include/fuse_core/util.h

```diff
@@ -340,6 +395,264 @@ static inline void quaternionToAngleAxis(const double * q, double * angle_axis,
   }
 }
 
+/**
+ * @brief Compute the jacobian of a quaternion normalization
+ *
+ * @param[in] q The quaternion
+ * @return      The jacobian of the quaternion normalization
+ */
+static inline fuse_core::Matrix4d jacobianQuatNormalization(Eigen::Quaterniond q)
+{
+  fuse_core::Matrix4d ret;
+  ret <<
+     q.x() * q.x() + q.y() * q.y() + q.z() * q.z(), -q.w() * q.x(), -q.w() * q.y(), -q.w() * q.z(),
+     -q.x() * q.w(), q.w() * q.w() + q.y() * q.y() + q.z() * q.z(), -q.x() * q.y(), -q.x() * q.z(),
+     -q.y() * q.w(), -q.y() * q.x(), q.w() * q.w() + q.x() * q.x() + q.z() * q.z(), -q.y() * q.z(),
+     -q.z() * q.w(), -q.z() * q.x(), -q.z() * q.y(), q.w() * q.w() + q.x() * q.x() + q.y() * q.y();
+
+  ret /= std::pow(q.norm(), 3.);
+  return ret;
+}
+
+/**
+ * @brief Convert a pose covariance in x, y, z, roll, pitch, yaw to a pose covariance in x, y, z, qw, qx, qy, qz
+ *
+ * @param[in] pose    The pose which covariance should be converted
+ * @param[in] cov_rpy The covariance in x, y, z, roll, pitch, yaw
+ * @return            The pose covariance in x, y, z, qw, qx, qy, qz
+ */
+static inline fuse_core::Matrix7d convertToPoseQuatCovariance(
+  const Eigen::Isometry3d& pose, const fuse_core::Matrix6d& cov_rpy)
+{
+  double rpy[3];
+  Eigen::Quaterniond q(pose.rotation());
+  double q_array[4] = {  // NOLINT(whitespace/braces)
+    q.w(),
+    q.x(),
+    q.y(),
+    q.z()
+  };
+  quaternion2rpy(q_array, rpy);
+
+  double J_rpy2quat[12];
+  rpy2quaternion(rpy, q_array, J_rpy2quat);
+
+  Eigen::Map<Eigen::Matrix<double, 4, 3, Eigen::RowMajor>> J_rpy2quat_map(J_rpy2quat);
+
+  Eigen::Matrix<double, 7, 6, Eigen::RowMajor> J;
+  J.topLeftCorner<3, 3>().setIdentity();
+  J.topRightCorner<3, 3>().setZero();
+  J.bottomLeftCorner<4, 3>().setZero();
+  J.bottomRightCorner<4, 3>() = J_rpy2quat_map;
+
+  return J * cov_rpy * J.transpose();
+}
+
+/**
+ * @brief Convert a pose covariance in x, y, z, qw, qx, qy, qz to a pose covariance in x, y, z, roll, pitch, yaw
+ *
+ * @param[in] pose    The pose which covariance should be converted
+ * @param[in] cov_rpy The covariance in x, y, z, qw, qx, qy, qz
+ * @return            The pose covariance in x, y, z, roll, pitch, yaw
+ */
+static inline fuse_core::Matrix6d convertToPoseRPYCovariance(
```

Comment:
Same as above it calcualtes the jacobian not the covariance

Reply:
Same as above, no, it does not just compute the jacobian for the pose but then uses the jacobian to convert the covariance represented in [x, y, z, qw, qx, qy, qz] to a covariance in [x, y, z, roll, pitch, yaw]. The conversion itself happens in the last line of the function.

Reply:
ah ok